com.unity.render-pipelines.high-definition/Runtime/Material/AxF/AxFData.hlsl

```diff
@@ -1,12 +1,428 @@
+//-------------------------------------------------------------------------------------
+// Defines
+//-------------------------------------------------------------------------------------
+// Gradients are now required:
+#define SURFACE_GRADIENT // Note: this affects Material/MaterialUtilities.hlsl's GetNormalWS() and makes it expect a surface gradient.
+
+//to test #define FLAKES_TILE_BEFORE_SCALE
+#define AXF_REUSE_SCREEN_DDXDDY
+// ...ie use _GRAD sampling for everything and calculate those only one time:
+// offset doesn't change derivatives, and scales just scales them, so we can cache them.
+
+// The compiler can't unroll the lightloop if flakes are sampled inside it, so we need to cache either LOD
+// or derivatives. We prefer the later, as the CalculateLevelOfDetail will not work when anisotropic filtering
+// is used, and AxF materials textures often have trilinear filtering set.
+#define FLAKES_USE_DDXDDY
+
+#define AXF_USES_RG_NORMAL_MAPS // else, RGB
+
 //-------------------------------------------------------------------------------------
 // Fill SurfaceData/Builtin data function
 //-------------------------------------------------------------------------------------
+#include "Packages/com.unity.render-pipelines.core/ShaderLibrary/Sampling/SampleUVMapping.hlsl"
 #include "Packages/com.unity.render-pipelines.high-definition/Runtime/Material/BuiltinUtilities.hlsl"
 #include "Packages/com.unity.render-pipelines.high-definition/Runtime/Material/MaterialUtilities.hlsl"
 #include "Packages/com.unity.render-pipelines.high-definition/Runtime/Material/Decal/DecalUtilities.hlsl"
 
+//-----------------------------------------------------------------------------
+// Texture Mapping
+//-----------------------------------------------------------------------------
+#ifdef AXF_USES_RG_NORMAL_MAPS
+#define AXF_DERIVATIVE_NORMAL UnpackDerivativeNormalRGorAG
+#else
+#define AXF_DERIVATIVE_NORMAL UnpackDerivativeNormalRGB
+#endif
+
 // Note: the scaling _Material_SO.xy should already be in texuv, but NOT the bias.
-#define AXF_TRANSFORM_TEXUV(texuv, name) ((texuv.xy) * name##_SO.xy + name##_SO.zw + _Material_SO.zw)
+#define AXF_TRANSFORM_TEXUV_BYNAME(texuv, name) ((texuv.xy) * name##_SO.xy + name##_SO.zw + _Material_SO.zw)
+#define AXF_GET_SINGE_SCALE_OFFSET(name) (name##_SO)
+#define AXF_TRANSFORM_TEXUV(texuv, scaleOffset) ((texuv.xy) * scaleOffset.xy + scaleOffset.zw + _Material_SO.zw)
+
+// Note: the scaling _Material_SO.xy should already be in ddx and ddy:
+#define AXF_SCALE_DDXDDY_BYNAME(vddx, name) ((vddx) * (name##_SO.xy))
+
+#if 0
+#define DDX(param) ddx_fine(param)
+#define DDY(param) ddy_fine(param)
+#else
+#define DDX(param) ddx(param)
+#define DDY(param) ddy(param)
+#endif
+
+struct TextureUVMapping
```

Comment:
Regarding your Macro system for texture and this code. Please don't duplicate code that already exist.
We have  SampleUVMapping.hlsl files that do similar thing.
If there is thing missing please update SampleUVMapping.hlsl or you can also clean up SampleUVMapping.hlsl to be more like your system. But they do the same thing (i.E handle surface gradient and all king of planar / triplanar.

If the UVMapping don't fit your need (which is the case as you need ddy/ddy
Maybe try to add a define (USE_GRADIENT) that enable it.  in UVMapping , intead of duplicting everything with TextureUVMapping. Etc...


Please avoid duplicating code or do your own system in isolation this make thing very hard to maintain. In particular if someone new look at the code.